File: sdk/js/src/Stream.js
import { EventEmitter } from 'events';
import VideoElement from './VideoElement';

const VideoResolutions =
{
    qvga: { width: { ideal: 320 }, height: { ideal: 180 } },
    vga: { width: { ideal: 640 }, height: { ideal: 360 } },
    shd: { width: { ideal: 960 }, height: { ideal: 540 } },
    hd: { width: { ideal: 1280 }, height: { ideal: 720 } }
};

export default class Stream extends EventEmitter {

    constructor(mid = null, stream = null) {
        super();
        this._mid = mid;
        this._stream = stream;
        this._videoElement = new VideoElement();
    }

    async init(sender = false, options = { audio: true, video: true, screen: false, resolution: 'hd' }) {
        if (sender) {
            if (options.screen) {
                this._stream = await navigator.mediaDevices.getDisplayMedia({ video: true });
            } else {
                this._stream = await navigator.mediaDevices.getUserMedia(
                    {
                        audio: options.audio,
                        video: options.video === true ?
                            VideoResolutions[options.resolution] : false
                    }
                );
            }
        }
    }

    set mid(id) { this._mid = id; }

    get mid() { return this._mid; }

    get stream() { return this._stream };

    render(elementId) {
        this._videoElement.play({ id: this._mid, stream: this._stream, elementId });
    }

    async stop() {
        this._videoElement.stop();
    }
}

File: sdk/flutter/lib/src/stream.dart
import 'package:events2/events2.dart';
import 'package:flutter_webrtc/webrtc.dart';
import 'logger.dart' show Logger;

// 'vga': {'minWidth': '640', 'minHeight': '360'},
final Map<String, dynamic> resolutions = {
  'qvga': {'width': '320', 'height': '180'},
  'vga': {'width': '640', 'height': '360'},
  'shd': {'width': '960', 'height': '540'},
  'hd': {'width': '1280', 'height': '720'},
  'fullhd': {'width': '1920', 'height': '1080'}
};

class Stream extends EventEmitter {
  var logger = new Logger("Ion::Stream");
  String _mid;
  MediaStream _stream;
  Stream([this._mid, this._stream]);

  init(
      [sender = false,
      audio = true,
      video = true,
      screen = false,
      quality = 'hd']) async {
    if (sender) {
      if (screen) {
        this._stream = await navigator
            .getDisplayMedia(_buildMediaConstraints(false, true));
      } else {
        Map<String, dynamic> videoConstrains = {
          "mandatory": {
            "minWidth": resolutions[quality]['width'],
            "minHeight": resolutions[quality]['height'],
            "minFrameRate": '30',
          },
          "facingMode": 'user',
        };
        this._stream = await navigator
            .getUserMedia(_buildMediaConstraints(audio, videoConstrains));
      }
    }
  }

  set mid(id) {
    this._mid = id;
  }

  String get mid => _mid;

  MediaStream get stream => _stream;

  _buildMediaConstraints(audio, video) {
    return {"audio": audio, "video": video ?? false};
  }
}
